### PR TITLE
Authorization + heroes claim system. DM has all rights

### PR DESCRIPTION
**Pull request by Darkoboar**

### CODE REVIEW

File: d20potz.py [hunk 1]
```diff
@@ -53,6 +53,8 @@ def read_cards(cards_dir):
 CONFIG = read_configuration("./d20potz.cfg", "./default.cfg")
 CARDS = read_cards(CONFIG.cards_dir)
 DB = leveldb.LevelDB(CONFIG.db_location)
+USERNAMES = dict([(5276124455, "Kostya"), (112553360, "Maxim"), (5225826508, "Vitalik"), (42026173, "Fedir"), (470866600, "Alex")])
```
Comment (vit-ka) [inline d20potz.py]:
```suggestion
USERNAMES = {5276124455:, "Kostya", 112553360: "Maxim"), (5225826508, "Vitalik"), (42026173, "Fedir"), (470866600, "Alex")])
```

Comment (bestander) [reply d20potz.py]: Хуестя, давай уж без хардкода

File: d20potz.py [hunk 2]
```diff
@@ -480,6 +562,59 @@ async def cards_command(update: Update, context: ContextTypes.DEFAULT_TYPE):
...
+            await context.bot.send_message(
+                chat_id=chat_id, text="{} claimed {}".format(to_username(user_id), player_name)
+            )
+            set_claim_status(chat_id, player_name, user_id, True)
+        else:
+            await context.bot.send_message(
+                chat_id=chat_id, text="You've already claimed {}, greedy potz".format(current_claim)
+            )
+
+async def unclaim_command(update: Update, context: ContextTypes.DEFAULT_TYPE):
```
Comment (bestander) [inline d20potz.py]: unnecessary command, claim should just over-wright the previous owner

File: d20potz.py [hunk 3]
```diff
@@ -480,6 +562,59 @@ async def cards_command(update: Update, context: ContextTypes.DEFAULT_TYPE):
...
+        )
+        return
+
+    player_name = params[1].lower()
+    if not await check_player_name(context, chat_id, player_name):
+        return   
+    
+    if is_player_claimed(chat_id, player_name):
+        if is_player_claimed_by_user(chat_id, player_name, user_id):
+            await context.bot.send_message(
```
Comment (bestander) [inline d20potz.py:528]: The logic is good but I think it's too strict.
Just allow claiming another character without unclaiming or checking if it was claimed already, it should be simpler.

a -> /claim heskan -> A claimed Heskan
b -> /claim vistra -> B claimed Vistra
a -> /claim vistra -> A claimed Vistra (Heskan is unclaimed and B does not own Vistra anymore)

File: d20potz.py (hunk 1 above)
Comment (nightshadows) [reply d20potz.py]: please remove the PII from the public repo. And delete this conversation

File: d20potz.py [hunk 4]
```diff
@@ -215,20 +255,18 @@ async def turn_command(update: Update, context: ContextTypes.DEFAULT_TYPE):
...
-    if player_name not in player_list:
-        await context.bot.send_message(
-            chat_id=chat_id,
-            text="{} is not one of {}".format(
-                spell_hero_name(player_name), player_list
-            ),
-        )
-        return
+    
+    if last_param in player_list:
```
Comment (nightshadows) [inline d20potz.py:265]: this will not work for "Vistra" for example

File: d20potz.py [hunk 5]
```diff
@@ -479,25 +508,64 @@ async def cards_command(update: Update, context: ContextTypes.DEFAULT_TYPE):
...
+        return
+
+    player_name = params[1].lower()
+    if not await check_player_name(context, chat_id, player_name):
+        return   
+    
+    if is_player_claimed(chat_id, player_name):
+        if is_player_claimed_by_user(chat_id, player_name, user_id):
+            await context.bot.send_message(
+                chat_id=chat_id, text="You've already claimed {}, stupid potz".format(player_name)
```
Comment (nightshadows) [inline d20potz.py]: lets not use 'potz' in this context

File: d20potz.py (hunk 3 above)
Comment (nightshadows) [reply d20potz.py:528]: +

**Review by nightshadows — CHANGES_REQUESTED** [review]
NO PII IN THE CODE. Make sure that this is NOT in the public repos after the review is closed

File: d20potz.py [hunk 6]
```diff
@@ -169,7 +169,47 @@ async def send_cards(
     if media_list:
         await context.bot.send_media_group(chat_id=chat_id, media=media_list)
 
+async def check_player_name(context: ContextTypes.DEFAULT_TYPE, chat_id, player_name):
+    player_list = CONFIG.order.lower().split()
+    if player_name not in player_list:
```
Comment (vit-ka) [inline d20potz.py:174]: :love:

File: d20potz.py [hunk 7]
```diff
@@ -169,7 +169,47 @@ async def send_cards(
...
+    else:
+        DB.Delete(claim_key)
 
+def get_player_by_user(chat_id, user_id):
+    player_list = CONFIG.order.lower().split()
+    
+    for player_name in player_list:
+        claim_key = f"player_claim_{chat_id}_{player_name}".encode("utf-8")
+        claimant = DB.Get(claim_key, default = None)
+        if claimant != None and claimant.decode() == str(user_id):
```
Comment (vit-ka) [inline d20potz.py:198]:
```suggestion
        if claimant is not None and claimant.decode() == str(user_id):
```

Comment (vit-ka) [reply d20potz.py:198]: Kinda similar to java equals(). A class can override `== None`, but not `is None`

**Review by vit-ka — APPROVED** [review]

**Pull request merged**